packages/format-library/src/link/inline.js

```diff
@@ -56,181 +64,55 @@ const URLPopoverAtLink = ( { isActive, addingLink, value, ...props } ) => {
-		this.setLinkTarget = this.setLinkTarget.bind( this );
-		this.onFocusOutside = this.onFocusOutside.bind( this );
-		this.resetState = this.resetState.bind( this );
-		this.autocompleteRef = createRef();
-
-		this.state = {
-			opensInNewWindow: false,
-			inputValue: '',
-		};
-	}
-
-	static getDerivedStateFromProps( props, state ) {
-		const { activeAttributes: { url, target } } = props;
-		const opensInNewWindow = target === '_blank';
-
-		if ( ! isShowingInput( props, state ) ) {
-			const update = {};
-			if ( url !== state.inputValue ) {
-				update.inputValue = url;
-			}
-
-			if ( opensInNewWindow !== state.opensInNewWindow ) {
-				update.opensInNewWindow = opensInNewWindow;
-			}
-			return Object.keys( update ).length ? update : null;
-		}
-
+	if ( ! isActive && ! addingLink ) {
 		return null;
 	}
 
-	onKeyDown( event ) {
-		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
-			// Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
-			event.stopPropagation();
-		}
-	}
-
-	onChangeInputValue( inputValue ) {
-		this.setState( { inputValue } );
-	}
-
-	setLinkTarget( opensInNewWindow ) {
-		const { activeAttributes: { url = '' }, value, onChange } = this.props;
-
-		this.setState( { opensInNewWindow } );
-
-		// Apply now if URL is not being edited.
-		if ( ! isShowingInput( this.props, this.state ) ) {
-			const selectedText = getTextContent( slice( value ) );
-
-			onChange( applyFormat( value, createLinkFormat( {
-				url,
-				opensInNewWindow,
-				text: selectedText,
-			} ) ) );
-		}
-	}
-
-	editLink( event ) {
-		this.setState( { editLink: true } );
-		event.preventDefault();
-	}
-
-	submitLink( event ) {
-		const { isActive, value, onChange, speak } = this.props;
-		const { inputValue, opensInNewWindow } = this.state;
-		const url = prependHTTP( inputValue );
+	const linkValue = {
+		url: activeAttributes.url,
+		opensInNewTab: activeAttributes.target === '_blank',
+	};
+	const onChangeLink = ( newValue ) => {
+		const newUrl = prependHTTP( newValue.url );
 		const selectedText = getTextContent( slice( value ) );
 		const format = createLinkFormat( {
-			url,
-			opensInNewWindow,
+			url: newUrl,
+			opensInNewWindow: newValue.opensInNewTab,
 			text: selectedText,
 		} );
 
-		event.preventDefault();
-
 		if ( isCollapsed( value ) && ! isActive ) {
-			const toInsert = applyFormat( create( { text: url } ), format, 0, url.length );
+			const toInsert = applyFormat( create( { text: newUrl } ), format, 0, newUrl.length );
 			onChange( insert( value, toInsert ) );
 		} else {
 			onChange( applyFormat( value, format ) );
 		}
 
-		this.resetState();
-
-		if ( ! isValidHref( url ) ) {
+		if ( ! isValidHref( newUrl ) ) {
 			speak( __( 'Warning: the link has been inserted but may have errors. Please test it.' ), 'assertive' );
 		} else if ( isActive ) {
 			speak( __( 'Link edited.' ), 'assertive' );
 		} else {
 			speak( __( 'Link inserted.' ), 'assertive' );
 		}
```

Comment:
Not sure if you have any thoughts on: What generalized behaviors can we extract around communicating link selection to the user.

There are a few issues I see with how this is currently implemented:

- The handling is split, where [`URLInput` will announce](https://github.com/WordPress/gutenberg/blob/adf6735f85858eb83e56bdcb7d21c25c6eee9742/packages/block-editor/src/components/url-input/index.js#L211-L212) when a suggestion link is "selected" using the <kbd>Tab</kbd> key, but then the application of that link is implemented in the submission handler
- Problem: If `URLInput` is supposed to behave like a raw input field, what should happen when <kbd>Enter</kbd> is pressed and there is no submission handler (i.e. a default/fallback behavior)?
- Case in point: There is no handler that communicates selection in either the Navigation block or Buttons block link editors. Therefore, the user is not informed when a selection is made.

It's difficult, because I think the message might need to change depending on where it's used (like here, "Link inserted" based on the changes applied in the paragraph).

A possible compromise solution:

- Always announce "Link selected" in URLInput for _both_ <kbd>Tab</kbd> and <kbd>Enter</kbd> handling
- Either:
   a. If that's sufficient information to indicate that a link can be assumed to have been applied, then remove this custom handling. I don't really think we want to make this assumption, considering that there are many other ways that a link can be applied (_clicking a suggestion_, tabbing to the "Submit" button and pressing Enter, etc).
   b. Otherwise, be more consistent in how we implement spoken messages for communicating how those selections are used. In other words, implement `speak` in the Navigation/Button blocks' [`onChange` handling](https://github.com/WordPress/gutenberg/blob/adf6735f85858eb83e56bdcb7d21c25c6eee9742/packages/block-library/src/navigation-link/edit.js#L170-L179) (or `onClose` ? Not sure when it should be considered as "persisted").

It might risk over-communicating though, since in many cases it would announce two messages of "Link selected", "Link inserted" when pressing <kbd>Enter</kbd>.

Related: https://github.com/WordPress/gutenberg/pull/10838#issuecomment-573774362